Enhanced Gyroscope Controls, Physics, and Tiltables

camblomquist opened the pull request:
As before, this branch aims to create a better control scheme than the accelerometer approach. With the accelerometer, a push of the phone would be interpreted as a tilt and there is no easy way to differentiate the two. Additionally, the previous implementation did not have any way to handle different starting rotations. The choice to manipulate the gravity of the scene instead of the game board was made so that resetting the phone to a neutral position would be trivial. Resetting the position and rotation of the game board itself would require work to ensure that physics objects maintained their relative position on the board.

This approach takes the gyroscope and stores a neutral position, then all physics are applied as a rotation from that neutral position. The issues with the previous implementation of this enhanced gyro have been fixed, including the bad light tilting and camera tilt as well as the snapping to an angle depending on the neutral position. The Camera script still needs a bit of work to implement a dead zone. where the camera doesn't move while the player remains within this zone.

I've tested this using Unity Remote on iOS and on an Android development build, both work as expected.

Any complaints from Unity about a skybox should be ignored.

New to this branch is also the virtual gyroscope, which allows for gyroscope input in the editor without Remote.

Close #24, close #33

Pierre-Demessence committed:
Version 0.1 - Prototype

conorlawton committed:
Renamed Camera script

Renamed Camera script to prevent conflict with Camera component.
Made Y Offset public and set to 20.

camblomquist committed:
Implement Enhanced Gyro Controls

New GravityController (rename to GyroController?) directly modifies the gravity of the world. A method has been added to reset the gyroscope so that the gravity vector interprets the current attitude as a resting position. 

Since the maze no longer moves, the directional light rotates with the gyroscope. This specific behavior will need modification.

conorlawton committed:
Editted ball physics

Added physics material to Ball.
Increased Ball mass to 100.
Fixed gyro tilting clamp issue.

conorlawton committed:
Actually fixed the clamp issue

conorlawton committed:
Updated .gitignore

Ignoring ProBuilder asset changes

conorlawton committed:
Merge branch 'feature/maze' into feature/enhanced-gyro

conorlawton committed:
Removed rigidbody on maze

Pierre-Demessence committed:
Fixed missing scene

camblomquist committed:
Implemented Tiltable and TiltableCamera

Tiltable and TiltableCamera provide various methods of tilting an object based on the gyroscope. Both components will tilt an object as a factor of the gyroscope and will tilt to a maximum angle accordingly. For example, A Tiltable with a maximum angle of 45 but a tilt factor of 0.5 will require a 90 tilt of the gyroscope to reach its peak. Any axis of rotation can be constrained. A Tiltable can choose to orbit an object instead of rotating about its origin. TiltableCamera works the same as the Tiltable orbit mode, but executes in LateUpdate as a matter of convention. These rotations are applied on top of the initial rotation of the object.

GravityController has been renamed to a more generic "Gyro" and it now exposes the modified tilt of the gyroscope for other classes to use. 
Angles have finally been fixed. Quaternions had a problem using signed angles, but the math performed on the tilt is easier to implement using signed angles. Now when given an Euler Angle, the methods first convert to an signed angle and then return the unsigned angle after modification.

Assorted test objects have been added, including a helix and bumper bowl. The bumper implemented is separate from the one implemented in feature/traps.

General organizational changes. Physics Materials are now a separate folder at the root instead of being a sub-folder of Materials. Scripts pertaining to Gyro controls are now in sub-folder Gyro and scripts that are defined as Modifiers (sort of like Decorators,) are in the sub-folder Modifiers. Editor scripts have been created for the new gyro scripts.

camblomquist committed:
Merge remote-tracking branch 'origin/develop' into feature/enhanced-gyro

camblomquist committed:
Cleanup

Post-merge cleanup. Renamed the "Script" objects in the scene to Manager and moved them to origin. Manager contains the Ctrl scripts as before and the one in Game contains Gyro as well.

camblomquist committed:
Fix Non-Development Build Bug

The non-development build code referenced a variable that is only compiled in development builds and the editor. This has been changed.

Pierre-Demessence committed:
Fixed missing reference

Pierre-Demessence committed:
Fixed collision sound pitch and volume formula so it depends on mass

Pierre-Demessence commented:
It works great but there's an issue when testing on my phone (real build, not Unity Remote) :

You press "Play", the gyro doesn't work.
You have to pause, go back to the menu and press play again for the gyro to work and anything to move.

camblomquist committed:
Implement Virtual Gyroscope

A custom editor window that allows you to control the Gyro without a gyroscope. Accessible through Tools > Virtual Gyro. You can drag the mouse in the model view like in the model inspector, this will change pitch and roll. There is a bug with weird rotations when manually adjusting pitch with the slider past 90 degrees in either direction. This shouldn't be an issue for this project though. Currently, there is no way to adjust Yaw in the model view, only with the slider.

camblomquist committed:
Merge remote-tracking branch 'origin/feature/enhanced-gyro' into feature/enhanced-gyro

Pierre-Demessence commented:
Update related to the last comment :
I enabled Dev Mode to test and here are the values :

> Gyro enabled : true
Neutral : 0;0;0;0
Attitude : [change depending on phone orientation]
Gravity : 0;-15;0 (doesn't change)
Pure : 0;0;0
Rotation : 0;0;0

Pierre-Demessence committed:
Added whole .idea folder in .gitignore

camblomquist committed:
Fix Invalid Gyroscope

More often than not, we would attempt to store a value from the gyroscope before it was initialized to a valid quaternion. The initial fix checked for equality to a zeroed value, but that solution didn't work with how Unity checks for equality. Now each component of the quaternion is checked.

camblomquist committed:
Merge remote-tracking branch 'origin/feature/enhanced-gyro' into feature/enhanced-gyro

Pierre-Demessence commented:
With the latest commit (8c0b995) when launching the game the camera and maze move corectly but the ball doesn't. Pressing the "Reset Gyro" button does not do anything anymore.

(So it seems the original issue was fixed but a new on appeared).

camblomquist committed:
Bugfixes

Hopefully squashed the bug with gyroscope input by moving most of the logic into Awake and setting timeScale to zero if a valid gyroscope input wasn't obtained. Start is now only in charge of waiting for that valid quaternion and setting the timeScale back to 1 if the quaternion wasn't valid on Start.

VirtualGyro had an issue with immediately jumping back to the previous position after resetting the orientation. This has been fixed by changing the drag value to zero when reset. The check for lastDir was removed as it turned out to not mean anything.

Pierre-Demessence committed:
Added "Reset Gyro" button to the Pause menu

Pierre-Demessence reviewed (dismissed):
Checked with the last commit

camblomquist committed:
Merge remote-tracking branch 'origin/develop' into feature/enhanced-gyro

Fixing merge conflicts

Pierre-Demessence reviewed (dismissed):
Good with the fixed conflicts

Pierre-Demessence committed:
Fixed previous merge.

Re-added victory screen buttons, re-added missing references on the GameCtrl as well as on the Goal.

Pierre-Demessence committed:
Removed now useless Console.Log

Now that we have a victory screen

Pierre-Demessence reviewed (dismissed):
Again more fixes

camblomquist committed:
Fix VirtualGyro Sliders

Fixed an issue where the sliders could not be used. Also fixed the problem with pitch not moving past 90 degrees in either direction.
